examples/control-vector-generator/control-vector-generator.cpp

```diff
@@ -312,21 +321,33 @@ static void padding_seq(llama_context * ctx, std::vector<llama_token> & tokens,
     }
 }
 
-static bool is_row_all_zeros(float * diff, size_t row, size_t cols, float eps = 1e-6) {
-    for (size_t i = 0; i < cols; ++i) if (diff[row * cols + i] > eps) return false;
+static bool is_row_all_zeros(struct ggml_tensor * diff, size_t row, size_t cols, float eps = 1e-6) {
+    for (size_t i = 0; i < cols; ++i) if (ggml_get_f32_nd(diff, row, i, 0, 0) > eps) return false;
     return true;
 }
 
 static void calc_diff(callback_data & cb_data) {
     // TODO: assert cb_data.v_pos.size() == cb_data.v_neg.size()
-    const size_t n_elems = cb_data.n_embd * cb_data.n_tokens;
     cb_data.v_diffs_wrapped.resize(cb_data.v_pos.size());
     for (size_t il = 0; il < cb_data.v_pos.size(); il++) {
         auto & inp_pos = cb_data.v_pos[il];
         auto & inp_neg = cb_data.v_neg[il];
-        float * dest = (float *) malloc(n_elems * sizeof(float));
-        for (size_t i = 0; i < n_elems; i++) {
-            dest[i] = inp_pos[i] - inp_neg[i];
+        auto n_bytes = ggml_nbytes(inp_pos);
+
+        struct ggml_init_params params = {
+            /*.mem_size   =*/ n_bytes,
+            /*.mem_buffer =*/ NULL,
+            /*.no_alloc   =*/ false,
+        };
+        struct ggml_context * ctx_data = ggml_init(params);
+
+        // TODO is this the best way to get dimension? i don't know which way n_embd/n_tokens go
+        // for that matter can we get rid of n_embd/n_tokens fields in favor of ne[0]/ne[1]?
+        struct ggml_tensor * dest = ggml_new_tensor_2d(ctx_data, GGML_TYPE_F32, inp_pos->ne[0], inp_pos->ne[1]);
+        for (size_t i = 0; i < cb_data.n_embd; i++) {
+            for (size_t j = 0; j < cb_data.n_tokens; j++) {
+                ggml_set_f32_nd(dest, i, j, 0, 0, ggml_get_f32_nd(inp_pos, i, j, 0, 0) - ggml_get_f32_nd(inp_neg, i, j, 0, 0));
```

Comment:
Because dest is only used internally to determine the differences before eliminating zero rows, is it necessary to use a ggml_tensor or would a float * suffice? I suppose the easy indexing provides easier readability than in a flattened float *, but is the overhead unnecessary?

Reply:
IMO `ggml_tensor` make it more clear to look at. In fact, here the `tensor->data` is just a `float *`, and `ggml_set_f32_nd` / `ggml_get_f32_nd` is simply equivalent to `tensor->data[...]`

Reply:
Sorry I've just looked again. Yes it seems like here it's better to just use `float *`, since this is just a temporary calculation